File: app/src/main/java/com/example/retrofitmvvmsample/Adapters/UserAdapter.java
package com.example.retrofitmvvmsample.Adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.retrofitmvvmsample.databinding.UsersListItemBinding;
import com.example.retrofitmvvmsample.modelClass.Datum;
import com.example.retrofitmvvmsample.utils.Utility;
import com.example.retrofitmvvmsample.viewHolder.UsersViewHolder;

import java.util.ArrayList;
import java.util.List;

public class UserAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {


    private final List<Datum> mListItems = new ArrayList<>();
    private final Context mContext;

    //constructor
    public UserAdapter(Context context) {
        this.mContext = context;
    }


    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        UsersListItemBinding binding = UsersListItemBinding
                .inflate(inflater, parent, false);
        return new UsersViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        UsersViewHolder userViewHolder;

        if (holder instanceof UsersViewHolder) {
            userViewHolder = (UsersViewHolder) holder;
            setUserViewHolder(userViewHolder, position);
        }
    }

    private void setUserViewHolder(UsersViewHolder userViewHolder, int position) {
        Datum user = mListItems.get(position);
        if (user != null) {
            userViewHolder.bind(user);
            Utility.loadImageUsingGlide(mContext,
                    userViewHolder.getBinding().userImageIv,
                    user.getAvatar());
        }
    }

    @Override
    public int getItemCount() {
        return mListItems.size();
    }

    //-------------------------------------Manipulating RecyclerView--------------------------------//
    public void clearData() {
        if (!mListItems.isEmpty()) {
            mListItems.clear();
            notifyDataSetChanged();
        }
    }

    public void addItemRange(List<Datum> items) {
        if (items != null) {
            int position = mListItems.size();
            mListItems.addAll(position, items);
            notifyItemRangeInserted(position, items.size());
        }
    }
}

File: app/src/main/java/com/example/retrofitmvvmsample/adapters/UserAdapter.kt
package com.example.retrofitmvvmsample.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import com.example.retrofitmvvmsample.databinding.UsersListItemBinding
import com.example.retrofitmvvmsample.modelClass.Datum
import com.example.retrofitmvvmsample.viewHolder.UsersViewHolder

class UserAdapter(private val mContext: Context) : ListAdapter<Datum, ViewHolder>(DiffCallback()) {

    class DiffCallback : DiffUtil.ItemCallback<Datum>() {
        override fun areItemsTheSame(oldItem: Datum, newItem: Datum): Boolean =
            oldItem.id == newItem.id

        override fun areContentsTheSame(oldItem: Datum, newItem: Datum): Boolean =
            oldItem.equals(newItem)

    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val binding = UsersListItemBinding
            .inflate(inflater, parent, false)
        return UsersViewHolder(binding, mContext)
    }

    override fun onBindViewHolder(
        holder: ViewHolder, position: Int) { //populate the data into the list_item (View Holder), as we scroll
        (holder as? UsersViewHolder)?.bindData(getItem(position))
    }
}